### DISCUSSION TITLE
tasmotised sonoff pow elite (pow320d) connected to aws iot core

### DISCUSSION THREAD

**Discussion by pauldegoede** [arendst/Tasmota#20851] [General]
Hi,

I'm struggling to get my device talking to AWS IoT core mqqt. The pow320d is a ESP32 device. I installed the default tasmota32.bin firmware.

The device is working and I can configure it / turn on/off via the web frontend. However I'm trying to connect it to aws iot core.

I have loaded the cloudformation template and I can push and receive messsages via MQTT client MQTTX. Now I just need to connect the tasmota device. I've got the private key file and the cert so need to get them loaded in TLSKey1 and TLSKey2.

I ran this command:
 Backlog MqttHost a1om9dt3a3e7rx-ats.iot.us-west-1.amazonaws.com; MqttPort 8883
And got this in the console:
09:59:21.612 MQT: Connect failed to a1om9dt3a3e7rx-ats.iot.us-west-1.amazonaws.com:8883, rc -4. Retry in 30 sec
09:59:52.445 MQT: Attempting connection...
09:59:56.820 MQT: Connect failed to a1om9dt3a3e7rx-ats.iot.us-west-1.amazonaws.com:8883, rc -4. Retry in 40 sec
10:00:37.671 MQT: Attempting connection...
10:00:42.047 MQT: Connect failed to a1om9dt3a3e7rx-ats.iot.us-west-1.amazonaws.com:8883, rc -4. Retry in 50 sec
10:01:32.895 MQT: Attempting connection...
10:01:37.310 MQT: Connect failed to a1om9dt3a3e7rx-ats.iot.us-west-1.amazonaws.com:8883, rc -4. Retry in 60 sec
10:02:38.115 MQT: Attempting connection...
10:02:42.497 MQT: Connect failed to a1om9dt3a3e7rx-ats.iot.us-west-1.amazonaws.com:8883, rc -4. Retry in 70 sec
10:03:53.349 MQT: Attempting connection...
10:03:57.710 MQT: Connect failed to a1om9dt3a3e7rx-ats.iot.us-west-1.amazonaws.com:8883, rc -4. Retry in 80 sec

And when I try and set TLSKey1 or TLSKey2 I get this:
09:58:56.526 CMD: TLSKey1
09:58:56.532 RSL: RESULT = {"Command":"Unknown"}
09:58:56.526 CMD: TLSKey2
09:58:56.532 RSL: RESULT = {"Command":"Unknown"}

Do I need to have a special build of tasmota32? My understanding is that tasmota32.bin has support for MQTT baked in (and there is a MQTT configure screen on the webui of the device.

And if you can help me with the values for TLSKey1 and TLSKey2 - in the article https://tasmota.github.io/docs/AWS-IoT-cert/#5-create-an-aws-iot-thing-with-private-key-and-certificate-once-per-tasmota-device step 5 it shows a short private key but mine is much longer. Is it fine to load the full private key?

**Comment by sfromis** [comment]
The meaning of "rc -4" is timeout. https://tasmota.github.io/docs/MQTT/#return-codes-rc
Maybe try a longer [MqttTimeout](https://tasmota.github.io/docs/Commands/#mqtttimeout)

And make sure that you have enabled TLS, [SetOption103](https://tasmota.github.io/docs/Commands/#setoption103)

**Reply by Noschvie** [reply]
@sfromis the MQTT socket timeout are seconds, isn't it?
Shall I add the unit to the docs?

**Comment by pauldegoede** [comment]
Hi sfromis,

Thanks for you reply.

TLS (SetOption103) is enable and the timeout is 30 seconds. Here's the output from the console:
17:08:43.848 MQT: Attempting connection...
17:08:45.516 MQT: TLS connection error: 0
17:08:45.517 MQT: Connect failed to a1om9dt3a3e7rx-ats.iot.us-west-1.amazonaws.com:8883, rc -2. Retry in 10 sec
17:08:47.922 QPC: Reset
17:08:56.342 MQT: Attempting connection...
17:08:58.000 MQT: TLS connection error: 0
17:08:58.001 MQT: Connect failed to a1om9dt3a3e7rx-ats.iot.us-west-1.amazonaws.com:8883, rc -2. Retry in 20 sec
17:09:18.851 MQT: Attempting connection...
17:09:20.518 MQT: TLS connection error: 0
17:09:20.519 MQT: Connect failed to a1om9dt3a3e7rx-ats.iot.us-west-1.amazonaws.com:8883, rc -2. Retry in 30 sec
17:09:39.596 CMD: MqttTimeout
17:09:39.599 RSL: RESULT = {"MqttTimeout":30}
17:09:49.096 CMD: SetOption103
17:09:49.099 RSL: RESULT = {"SetOption103":"ON"}

**Reply by sfromis** [reply]
As the page I linked says, rc -2 is simply that the network connection failed. Very different from what you posted earlier, and a faster response. Look outside of Tasmota to find out why it cannot connect. Could be the other end not ready for a connection, or whatever.

**Comment by pauldegoede** [comment]
Hi sfromis,

Thanks for you input.

Just to go back to my previous post:
1. I can post to aws iot mqtt via mqttx (loading the CA cert, private key and client cert)
2. when I run CMD: TLSKey1 I get RSL: RESULT = {"Command":"Unknown"}
3. when I run Backlog MqttHost a1om9dt3a3e7rx-ats.iot.us-west-1.amazonaws.com; MqttPort 8883 it fails - almost certainly because I haven't set the TLS cert info.

So my initial question was:
1. why is CMD TLSKey1 command unknown - this seems strange - do I need a special build that supports setting TLSKey1 and TLSKey2? I'm currently using latest stable of tasmota32.bin
2. or am I doing something wrong with the above so it's not the firmware but my config which is wrong.

Thanks
Paul

**Reply by sfromis** [reply]
Not using TLS or AWS, I have no "guide info" about getting things to work; I merely tried to provide hints about basic things like meaning of return codes.

**Reply by s-hadinger** [reply]
@pauldegoede I advise to use password authentication as the preferred method, and use device certificate only if you really need it. Using device certificates requires to compile your own binary

**Comment by pauldegoede** [comment]
Thanks sfromis. Hoping to get some feedback from people in the community using aws iot core + TLS.

**Reply by barbudor** [reply]
As far as I can see, the documentation page [AWS-IoT](https://tasmota.github.io/docs/AWS-IoT/) is wrong and USE_MQTT_AWS_IOT_LIGHT is NOT included in tasmota32
Which is also confirmed by [BUILDS.md](https://github.com/arendst/Tasmota/blob/development/BUILDS.md)

So far you need to build your own binary

@s-hadinger may know why the page doc is wrong. It may have been removed from the configuration but the page not edited.

**Reply by s-hadinger** [reply]
Ok thanks. Even without `USE_MQTT_AWS_IOT_LIGHT` it will connect successfully to AWS IoT, but it does not support IOT Shadows.
